crates/ark/src/lsp/main_loop.rs

```diff
@@ -692,3 +677,210 @@ impl std::fmt::Debug for TraceKernelNotification<'_> {
+    // Deduplicate tasks by keeping only the last one for each URI. We use a
+    // `HashMap` so only the last insertion is retained. This is effectively a
+    // way of cancelling diagnostics tasks for outdated documents.
+    let batch: std::collections::HashMap<_, _> = batch
+        .into_iter()
+        .map(|task| (task.uri, task.state))
+        .collect();
+
+    let mut futures = FuturesUnordered::new();
+
+    for (uri, state) in batch {
+        futures.push(task::spawn_blocking(move || {
+            let _span = tracing::info_span!("diagnostics_refresh", uri = %uri).entered();
+
+            if let Some(document) = state.documents.get(&uri) {
+                // Special case testthat-specific behaviour. This is a simple
+                // stopgap approach that has some false positives (e.g. when we
+                // work on testthat itself the flag will always be true), but
+                // that shouldn't have much practical impact.
+                let mut doc = document.clone();
+                if Path::new(uri.path())
+                    .components()
+                    .any(|c| c.as_os_str() == "testthat")
+                {
+                    doc.testthat = true;
+                };
+
+                let diagnostics = generate_diagnostics(doc, state.clone());
+                Some(RefreshDiagnosticsResult {
+                    uri,
+                    diagnostics,
+                    version: document.version,
+                })
+            } else {
+                None
+            }
+        }));
+    }
+
+    // Publish results as they complete
+    while let Some(Ok(Some(result))) = futures.next().await {
+        publish_diagnostics(result.uri, result.diagnostics, result.version);
+    }
+}
+
+pub(crate) fn index_start(folders: Vec<String>, state: WorldState) {
+    INDEXER_QUEUE
+        .send(IndexerQueueTask::Indexer(IndexerTask::Start { folders }))
+        .unwrap_or_else(|err| lsp::log_error!("Failed to queue initial indexing: {err}"));
+
+    diagnostics_refresh_all(state);
+}
+
+pub(crate) fn index_update(uri: Url, document: Document, state: WorldState) {
+    INDEXER_QUEUE
+        .send(IndexerQueueTask::Indexer(IndexerTask::Update {
+            document,
+            uri: uri.clone(),
+        }))
+        .unwrap_or_else(|err| lsp::log_error!("Failed to queue index update: {err}"));
+
+    // Refresh all diagnostics since the indexer results for one file may affect
+    // other files
+    diagnostics_refresh_all(state);
+}
+
+pub(crate) fn diagnostics_refresh_all(state: WorldState) {
+    for (uri, _document) in state.documents.iter() {
+        INDEXER_QUEUE
+            .send(IndexerQueueTask::Diagnostics(RefreshDiagnosticsTask {
+                uri: uri.clone(),
+                state: state.clone(),
+            }))
+            .unwrap_or_else(|err| lsp::log_error!("Failed to queue diagnostics refresh: {err}"));
```

Comment:
I think some of this was there before, but there is more `clone()`ing of the `WorldState` than I expected in this code

It feels like we are cloning like crazy:
- Call site of `index_start()`
- Call site of `index_update()`, every change, really??
- Each document in `diagnostics_refresh_all()` gets its own copy
- Call site of `generate_diagnostics()`, so basically every change, I think this was here before

I know the intention of `WorldState` is that we can send it to other threads but it just seems like it's happening a lot here, is that fully intentional?

Reply:
It's intentional because once the state is managed by Salsa, there will be no cloning at all. If needed we could review things in the meantime by switching e.g. to `Arc` for clone-free sharing.